Add cnxml-to-html command to transform index.cnxml to html

karenc opened the pull request:
Given a collection directory, use ~cnx-db~ cnx-transforms `cnxml_to_full_html` to
transform all the `index.cnxml` to `index.cnxml.html`.

---

I don't know how people feel about `cnxml2html` vs `cnxml-to-html` or something else.  I'm open to suggestions.

Depends on https://github.com/openstax/cnx-db/pull/174 and https://github.com/openstax/cnx-archive/pull/625

**NOTE:** Remove the last commit :skull_and_crossbones:  before merge.

pumazi commented:
I favor `cnxml-to-html`. But I'd do a quick slack poll to see what others want. I find dashes to be more readable in this case.

pumazi commented on nebu/cli/cnxml2html.py:
```diff
@@ -0,0 +1,23 @@
+    index.cnxml.html"""
+
+    for root, dirs, files in os.walk(col_path):
+        if 'index.cnxml' in files:
```
Any chance we can stick with using `pathlib` around the codebase? The `<Path>.glob(...)` method will likely do the searching you are walking around for here.

karenc replied:
So even though the files are already returned in `os.walk`, you want me to use `.glob()` to search for it?  If you confirm that's really what you want, I can change it.  I don't particularly agree with it but I'm not strongly against it.

pumazi commented on requirements/main.txt:
```diff
@@ -1,5 +1,6 @@
 click
 cnx-litezip>=1.6.0
+cnx-db
```
💦

reedstrm commented:
(pretty sure that's the right tracking issue :-) )

karenc commented:
@pumazi and @BryanEliDimas: I've changed `cnxml2html` to `cnxml-to-html`.

The usage now looks like this:

```
Usage: neb cnxml-to-html [OPTIONS] COLLECTION_PATH

  Given a collection COLLECTION_PATH (downloaded using neb get or complete
  zip on the cnx site), transform all the index.cnxml to index.cnxml.html
```

@BryanEliDimas There is no way to add a help message for `COLLECTION_PATH` so I
just added more details to the help message for the command.

I've also mocked `cnxml_to_full_html` in the test so it's possible to assert
the number of times called.

@pumazi and @BryanEliDimas: Could you just have a look and see if everything else
looks good?

I'm going to work on moving the transforms code to cnx-transforms.

karenc commented:
The transforms code has been moved to cnx-transforms:

https://github.com/openstax/cnx-transforms/pull/1

karenc commented:
@pumazi I know what's wrong...  It's not a caching issue that travis can't pip install cnx-transforms.  cnx-transforms is not available for python 3 :man_facepalming: 

https://pypi.org/project/cnx-transforms/#files

![cnxtransforms](https://user-images.githubusercontent.com/153353/53165899-c74f3180-35cb-11e9-8d46-54123cc6f042.jpg)

pumazi commented:
Aha, ok I'll get right on this and get another release out!

pumazi commented:
cnx-transforms 1.0.1 is a python3 compatible release. That should resolve the testing issue.

karenc commented:
@pumazi This is great, thanks!  Could you merge this PR when you feel it's ready?

karenc commented:
@pumazi I've updated the commit message to change cnx-db to cnx-transforms.  Thanks for fixing the description! :D

karenc commented:
Just did another small change, since I'm not using context, I'm removing `@click.pass_context`:

```diff
diff --git a/nebu/cli/cnxml_to_html.py b/nebu/cli/cnxml_to_html.py
index 8d9fd10..f2f8a8b 100644
--- a/nebu/cli/cnxml_to_html.py
+++ b/nebu/cli/cnxml_to_html.py
@@ -9,8 +9,7 @@ from ._common import common_params
 @click.command()
 @common_params
 @click.argument('collection_path')
-@click.pass_context
-def cnxml_to_html(ctx, collection_path):
+def cnxml_to_html(collection_path):
     """Given a collection COLLECTION_PATH (downloaded using neb get or complete
     zip on the cnx site), transform all the index.cnxml to index.cnxml.html"""
 
```

pumazi commented on docs/changes.rst:
```diff
@@ -2,6 +2,11 @@
 Change Log
 ==========
 
+9.1.0
```
FYI, I'm going to be releasing nebuchadnezzar as `9.1.0b1` <-- Note the `b1`. This specifies this as a pre-release, thus pip won't install it on the Content Managers system without them explicitly requesting it.

This will enable us to complete the BakedPDF CLI parts of nebuchadnezzar without interrupting the Content Manager's existing installation. And it will also mean the cnx-automation tests won't fail while we are still in development.

karenc replied:
Oh I see!  That's handy. :+1:

BryanEliDimas commented on nebu/tests/cli/test_cnxml_to_html.py:
```diff
@@ -0,0 +1,63 @@
+        for path in index_cnxml_html:
+            with open(path) as f:
+                content = f.read()
+            assert '<html><body>transformed html</body></html>' == content
```
Isn't this just testing that the mocking mechanism worked? When you mocked `cnxml_to_full_html` you also set its return value and that's all this line is testing. In Rails, for example, I would've called `.and_call_original` to actually test that the function works, not just that it gets called as many times as we expect (which is good, btw, and thanks for adding that after I requested, but not sufficient). In other words, it seems to me like you're breaking rule # 1 of mocking: "Don't mock what you don't own." @karenc

karenc replied:
I thought that's what you wanted, that `cnxml_to_full_html` was indeed called.  I have thought about calling the original function.  To me, that's the difference between unit testing and integration testing.  By mocking `cnxml_to_full_html`, I'm doing unit testing.  This only tests the code that I wrote in neb and not outside of it.  It works as long as `cnxml_to_full_html` doesn't change its interface.

BryanEliDimas replied:
> To me, that's the difference between unit testing and integration testing

True. I guess I expected one unit test and one integration test. The unit test just tests your code and it wouldn't break if `cnxml_to_full_html` breaks or changes 👍, but the integration test would (break), as it should, and then we'd have to fix just that one test. But I suppose that does sound like an improvement or a bug fix, so this one deserves/deserved to get merged and I should open up a separate issue for adding an integration test. No problem.
